### ISSUE TITLE
Search functionality not available on Android auto despite passing the necessary extras

### ISSUE THREAD

**Issue by yuroyami** [androidx/media#645]
In my `MediaLibraryService`, here's my implementation of `onGetLibraryRoot` : 

```kotlin
        override fun onGetLibraryRoot(session: MediaLibrarySession, browser: MediaSession.ControllerInfo, params: LibraryParams?): ListenableFuture<LibraryResult<MediaItem>> {
            val rootExtras = Bundle()
            rootExtras.putBoolean(MediaConstants.BROWSER_SERVICE_EXTRAS_KEY_SEARCH_SUPPORTED, true)
            val libraryParams = LibraryParams.Builder().setExtras(rootExtras).build()
            return Futures.immediateFuture(LibraryResult.ofItem(rootItem, libraryParams))
        }
```

The code above should make the search functionality available on Android Auto, but the search button doesn't even show up. I tried to replace `MediaConstants.BROWSER_SERVICE_EXTRAS_KEY_SEARCH_SUPPORTED` with `"android.media.browse.SEARCH_SUPPORTED"` to make sure I am passing the correct key, but no avail. 

The same result is met on both the virtual desktop head unit and a real car head unit. Am I doing something wrong ?

**Comment by marcbaechinger** [comment]
There are a few things that come to mind:

First, I think this is supported out of the box without you having to manually insert the extra. You only have to make sure that the browser has the command `COMMAND_CODE_LIBRARY_SEARCH` available. If this is the case, Media3 will insert that extra property automatically. This is done [here in `MediaLibraryServiceLegacyStub.onGetRoot`](https://github.com/androidx/media/blob/release/libraries/session/src/main/java/androidx/media3/session/MediaLibraryServiceLegacyStub.java#L122-L126). Vice versa, I would expect if you don't have this command available, Media3 would override your value with `false`.

So you would have to make sure that the `COMMAND_CODE_LIBRARY_SEARCH` is available. If you do not override `Callback.onConnect()` this is included by default. If you do override and give back a different set of commands then please make sure to include this command.

Second, of course, there may be a bug in Media3 that somehow doesn't correctly propagate these extras to the legacy browser client. For this I created a unit test case in [`MediaBrowserCompatWithMediaLibraryServiceTest`](https://github.com/androidx/media/blob/release/libraries/test_session_current/src/androidTest/java/androidx/media3/session/MediaBrowserCompatWithMediaLibraryServiceTest.java) to verify whether this works as expected:

```java

  @Test
  public void getRoot_commandSearchAvailable_expectedBrowserRootExtraIsTrue() throws Exception {
    connectAndWait(/* connectionHints= */ Bundle.EMPTY);

    assertThat(
        browserCompat
            .getExtras()
            .getBoolean(BROWSER_SERVICE_EXTRAS_KEY_SEARCH_SUPPORTED, /* defaultValue= */ false))
        .isTrue();
  }
```

This test runs green without me manipulating the available commands in `onConnect()`, which is the expected result.

I also tested this by using t[he media controller test app](https://github.com/googlesamples/android-media-controller), that uses the `MediaBrowserCompat` of the legacy media1 library that Automotive is using as well. When I connect with this app to the session demo, I can see the extras in the root keys that I see with `browserCompat.getExtras()`. This seems to have the expected key in the extras by default without me changing something in the demo app.

![image](https://github.com/androidx/media/assets/584321/4e86adc2-0dda-4d00-b663-80ad9aa799bd)

So from what I see and tested this should work given the `COMMAND_CODE_LIBRARY_SEARCH` is available which is the case by default. 

Can you review your code and let me know whether you provide the required command?

**Comment by yuroyami** [comment]
@marcbaechinger Note that I've done my testing using an emulator but I also confirm the issue exists on a real car headunit. 

Now on to details: I made sure that the required command does indeed exist in the available session commands. Here's my `onConnect` implementation: 

```kotlin
override fun onConnect(session: MediaSession, controller: MediaSession.ControllerInfo): MediaSession.ConnectionResult {
            val superCall = super.onConnect(session, controller)

            val sessionComs = superCall.availableSessionCommands
                .buildUpon()
                .add(CUSTOM_COM_CUSTOM1)
                .add(CUSTOM_COM_CUSTOM2)
                .build()

            //search command code is 50005
            for (comm in sessionComs.commands) {
                Log.e("SessionCommand", "${comm.commandCode} - ${comm.customAction}")
            }
            return MediaSession.ConnectionResult.accept(sessionComs, superCall.availablePlayerCommands)
        }
```

As expected, all library commands appear in the logcat as well as the two custom commands. But no sign of any search button throughout the lifecycle of the Android Auto interface.

It gets weirder: when I loaded the media3 branch of UAMP, and apparently the search functionality did not appear on the car screen either. I assume that the UAMP is supposed to provide the search functionality as well. 

It gets even weirder: At some point when I started my media3 development, the search functionality did exist, then at some point it just stopped working. I think there is something else messing with it, probably another MediaLibraryCallback method ? I will keep on testing and see what I can fruit out of it.

**Comment by marcbaechinger** [comment]
When I [declare media support for Android Auto](https://developer.android.com/training/cars/media/auto#manifest-car-app) in the manifest of the session demo app app, then I see the search icon and I can search the media library as expected.

To make this work it needs 3 pieces: 

- On connect the Auto controller needs to have `COMMAND_CODE_LIBRARY_SEARCH` available. This is the case by default. 
-  `Callback.onSearch(...)` is implemented and returns `RESULT_SUCCESS` and calls `session.notifySearchResultChanged` to notify the browser about the number of matches.
- `Callback.onGetSearchresult(...)` is implemented and returns the actual media items of the search result.

With this in place this worked for me on Android Auto (HUD) and Automotive OS (Emulator with API 29).

**Comment by yuroyami** [comment]
It seems that i haven't yet gotten to a workaround or a solution, the source of the problem remains a mystery. I will fiddle around with my MediaLibraryService and update this thread.

**Comment by PawelTypiak** [comment]
I have the opposite problem - using `rootExtras.putBoolean(android.media.browse.SEARCH_SUPPORTED, false)` should hide the search button, but unfortunately it doesn't work on the 1.2.0 and 1.2.1 version. It works on UAMP Media3 demo app but it has 1.0.0-beta01 version of Media3. Is there any solution to hide the search button?

**Comment by marcbaechinger** [comment]
I'm closing this after testing again with tot that will be released with 1.4.1.

The search icon is available on AAOS emulator for API 29 and API 33 if `COMMAND_CODE_LIBRARY_SEARCH` is included in the available commands in `MediaSession.Callback.onConnect(MediaSession, ControllerInfo, ...)` when the controller of Automotive identified with  `MediaSession.isAutomotiveController(controllerInfo)` connects:

Disabling the search for Auto and Automotive (removing the search command from the default command set):

```kotlin
override fun onConnect(
    session: MediaSession,
    controller: MediaSession.ControllerInfo): MediaSession.ConnectionResult {
  if (session.isAutomotiveController(controller) ||
      session.isAutoCompanionController(controller)) {
    return MediaSession.ConnectionResult.AcceptedResultBuilder(session)
      .setAvailableSessionCommands(mediaNotificationSessionCommands)
      .build()
  }
 // all other controllers
 return MediaSession.ConnectionResult.AcceptedResultBuilder(session).build()
}
```
